File: crates/claw-models/src/cron_schedule.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Deserializer, Serialize};
use std::path::PathBuf;
use uuid::Uuid;

use crate::OutputDest;

/// Deserialize a Vec<String> that might be stored as {} (empty object) in Redis.
/// This happens when some code paths serialize empty collections as objects.
fn vec_string_or_empty<'de, D>(deserializer: D) -> Result<Vec<String>, D::Error>
where
    D: Deserializer<'de>,
{
    use serde::de;
    struct VecOrEmpty;
    impl<'de> de::Visitor<'de> for VecOrEmpty {
        type Value = Vec<String>;
        fn expecting(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
            f.write_str("a sequence or empty object")
        }
        fn visit_seq<A: de::SeqAccess<'de>>(self, mut seq: A) -> Result<Vec<String>, A::Error> {
            let mut v = Vec::new();
            while let Some(s) = seq.next_element()? {
                v.push(s);
            }
            Ok(v)
        }
        fn visit_map<A: de::MapAccess<'de>>(self, _map: A) -> Result<Vec<String>, A::Error> {
            Ok(Vec::new()) // {} → empty vec
        }
    }
    deserializer.deserialize_any(VecOrEmpty)
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CronSchedule {
    pub id: Uuid,
    pub name: String,
    pub schedule: String,
    pub enabled: bool,
    pub prompt: String,
    #[serde(default, deserialize_with = "vec_string_or_empty")]
    pub skill_ids: Vec<String>,
    #[serde(default = "default_cron_working_dir")]
    pub working_dir: PathBuf,
    pub model: Option<String>,
    pub max_budget_usd: Option<f64>,
    #[serde(default)]
    pub output_dest: OutputDest,
    #[serde(default, deserialize_with = "vec_string_or_empty")]
    pub tags: Vec<String>,
    #[serde(default = "default_cron_priority")]
    pub priority: u8,
    pub workspace_id: Option<Uuid>,
    pub template_id: Option<Uuid>,
    pub last_run: Option<DateTime<Utc>>,
    pub last_job_id: Option<Uuid>,
    pub created_at: DateTime<Utc>,
}

fn default_cron_working_dir() -> PathBuf {
    PathBuf::from(".")
}

fn default_cron_priority() -> u8 {
    5
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateCronRequest {
    pub name: String,
    pub schedule: String,
    #[serde(default = "default_enabled")]
    pub enabled: bool,
    pub prompt: String,
    #[serde(default, deserialize_with = "vec_string_or_empty")]
    pub skill_ids: Vec<String>,
    pub working_dir: Option<PathBuf>,
    pub model: Option<String>,
    pub max_budget_usd: Option<f64>,
    #[serde(default)]
    pub output_dest: OutputDest,
    #[serde(default, deserialize_with = "vec_string_or_empty")]
    pub tags: Vec<String>,
    pub priority: Option<u8>,
    pub workspace_id: Option<Uuid>,
    pub template_id: Option<Uuid>,
}

fn default_enabled() -> bool {
    true
}

File: flutter_ui/lib/models/cron_schedule.dart
class CronSchedule {
  final String id;
  final String name;
  final String schedule;
  final bool enabled;
  final String prompt;
  final List<String> skillIds;
  final List<String> toolIds;
  final String workingDir;
  final String? model;
  final double? maxBudgetUsd;
  final Map<String, dynamic>? outputDest;
  final List<String> tags;
  final int priority;
  final String? workspaceId;
  final String? templateId;
  final String? lastRun;
  final String? lastJobId;
  final String createdAt;

  CronSchedule({
    required this.id,
    required this.name,
    required this.schedule,
    this.enabled = true,
    required this.prompt,
    this.skillIds = const [],
    this.toolIds = const [],
    this.workingDir = '.',
    this.model,
    this.maxBudgetUsd,
    this.outputDest,
    this.tags = const [],
    this.priority = 5,
    this.workspaceId,
    this.templateId,
    this.lastRun,
    this.lastJobId,
    required this.createdAt,
  });

  factory CronSchedule.fromJson(Map<String, dynamic> json) => CronSchedule(
        id: json['id'] ?? '',
        name: json['name'] ?? '',
        schedule: json['schedule'] ?? '',
        enabled: json['enabled'] ?? true,
        prompt: json['prompt'] ?? '',
        skillIds: List<String>.from(json['skill_ids'] ?? []),
        toolIds: List<String>.from(json['tool_ids'] ?? []),
        workingDir: json['working_dir'] ?? '.',
        model: json['model'],
        maxBudgetUsd: (json['max_budget_usd'] as num?)?.toDouble(),
        outputDest: json['output_dest'] is Map<String, dynamic>
            ? json['output_dest']
            : null,
        tags: List<String>.from(json['tags'] ?? []),
        priority: json['priority'] ?? 5,
        workspaceId: json['workspace_id'],
        templateId: json['template_id'],
        lastRun: json['last_run'],
        lastJobId: json['last_job_id'],
        createdAt: json['created_at'] ?? '',
      );

  Map<String, dynamic> toCreateJson() => {
        'name': name,
        'schedule': schedule,
        'enabled': enabled,
        'prompt': prompt,
        if (skillIds.isNotEmpty) 'skill_ids': skillIds,
        if (toolIds.isNotEmpty) 'tool_ids': toolIds,
        if (workingDir != '.') 'working_dir': workingDir,
        if (model != null) 'model': model,
        if (maxBudgetUsd != null) 'max_budget_usd': maxBudgetUsd,
        if (outputDest != null) 'output_dest': outputDest,
        if (tags.isNotEmpty) 'tags': tags,
        'priority': priority,
      };
}
